### PR TITLE
refactor(sdk): System Metrics probing/monitoring architecture refactor

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-11246
Fixes WB-9748

Description
-----------
In this PR, we refactor the SDK System Monitoring. Our goal is to provide clean, pythonic APIs, extensive out-of-the-box capabilities, and easy extensibility.

High-level description of the chosen approach:

- The `SystemMonitor` class uses composition to accumulate `Metric`s of different kinds originating from `Asset`s of different kinds.
- We prefer composition over inheritance: base classes `Asset` and `Metric` are defined as `typing.Protocols`.
- The `Asset` class encapsulates everything relating to e.g. CPU, GPU, TPU, Network, I/O etc.
    - Provides a mechanism to define custom resources.
    - Use `typing.Protocol`’s to define and gently enforce interfaces.
    - A `MetricsMonitor` object is used to collect metrics with a customizable sampling aggregation.
    - What’s in a name: we call it “Asset” because it’s something of value if you/we are tracking some metrics associated with it.
- `Asset`s are auto-discovered at runtime.

Testing
-------
- Re-wrote `test_meta` (as `test_system_info`), `test_ipu` (`test_system_metric_ipu`), and `test_tpu` (`test_system_metric_tpu`),
- Introduced new unit tests in `test_system_monitor`, `test_system_metric_gpu`.

### CODE REVIEW

File: wandb/sdk/system/assets/ipu.py [hunk 1]
```diff
@@ -145,7 +150,7 @@ def __init__(
     @classmethod
     def is_available(cls) -> bool:
         try:
-            import gcipuinfo  # noqa
+            import gcipuinfo  # type: ignore  # noqa: F401
         except ImportError:
             return False
```
Comment (kptkin) [inline wandb/sdk/system/assets/ipu.py:155]: let's just check the `gcipuinfo` that we import in the beginning of the script?

File: wandb/sdk/internal/handler.py [hunk 2]
```diff
@@ -683,18 +686,18 @@ def handle_request_run_start(self, record: Record) -> None:
         self._respond_result(result)
 
     def handle_request_resume(self, record: Record) -> None:
-        if self._system_stats is not None:
-            logger.info("starting system metrics thread")
-            self._system_stats.start()
+        if self._system_monitor is not None:
+            logger.info("starting system metrics process")
```
Comment (kptkin) [inline wandb/sdk/internal/handler.py:690]: this is a thread currently, probably want to fix the log message

File: wandb/sdk/internal/handler.py [hunk 3]
```diff
@@ -683,18 +686,18 @@ def handle_request_run_start(self, record: Record) -> None:
...
         if self._track_time is not None:
             self._accumulate_time += time.time() - self._track_time
         self._track_time = time.time()
 
     def handle_request_pause(self, record: Record) -> None:
-        if self._system_stats is not None:
-            logger.info("stopping system metrics thread")
-            self._system_stats.shutdown()
+        if self._system_monitor is not None:
+            logger.info("stopping system metrics process")
```
Comment (kptkin) [inline wandb/sdk/internal/handler.py:699]: same as above

File: wandb/sdk/system/assets/interfaces.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    Takes care of collecting, sampling, serializing, and publishing a set of metrics.
+    """
+
+    def __init__(
+        self,
+        metrics: List[Metric],
+        interface: Interface,
+        settings: "SettingsStatic",
+        shutdown_event: synchronize.Event,
+        # start_time,
```
Comment (kptkin) [inline wandb/sdk/system/assets/interfaces.py:100]: let's remove this one...

File: wandb/sdk/system/assets/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                0.1,
+                settings._stats_sample_rate_seconds,
+            )
+        )  # seconds
+        # The number of samples to aggregate (e.g. average or compute max/min etc.)
+        # before publishing; defaults to 15; valid range: [1:30]
+        self.samples_to_aggregate: int = min(
+            30, max(1, settings._stats_samples_to_average)
+        )
+        # print(self.sampling_interval, self.samples_to_aggregate)
```
Comment (kptkin) [inline wandb/sdk/system/assets/interfaces.py:118]: let's remove this one

File: wandb/sdk/system/assets/ipu.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    def start(self) -> None:
+        self.metrics_monitor.start()
+
+    def finish(self) -> None:
+        self.metrics_monitor.finish()
+
+    def probe(self) -> dict:
+        num_devices = len(self.metrics[0]._gc_ipu_info.getDevices())  # type: ignore
+        return {"ipu": {"device_count": num_devices, "vendor": "Graphcore"}}
```
Comment (kptkin) [inline wandb/sdk/system/assets/ipu.py:167]: we probably can make it more meaningful given the comment above that some metrics are static?

File: wandb/sdk/system/assets/network.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
+import multiprocessing as mp
+from collections import deque
+from typing import TYPE_CHECKING, List
+
+import psutil
```
Comment (kptkin) [inline wandb/sdk/system/assets/network.py:5]: should we fix everywhere making `psutil` conditional import?

File: wandb/sdk/system/system_monitor.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        if self.asset_interface is None:
+            return None
+        # only extract as many items as are available in the queue at the moment
+        size = self.asset_interface.metrics_queue.qsize()
+        # print(f"WOKE UP, FELL OUT OF BED, DRAGGED A COMB ACROSS MY HEAD: {size}")
+
+        serialized_metrics = {}
+        for _ in range(size):
+            item = self.asset_interface.metrics_queue.get()
+            # print(f"::harvested:: {item}")
```
Comment (kptkin) [inline wandb/sdk/system/system_monitor.py:102]: clean this one

File: wandb/sdk/system/system_monitor.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        self.system_info: SystemInfo = SystemInfo(
+            settings=settings, interface=interface
+        )
+
+    def aggregate_and_publish_asset_metrics(self) -> None:
+        if self.asset_interface is None:
+            return None
+        # only extract as many items as are available in the queue at the moment
+        size = self.asset_interface.metrics_queue.qsize()
+        # print(f"WOKE UP, FELL OUT OF BED, DRAGGED A COMB ACROSS MY HEAD: {size}")
```
Comment (kptkin) [inline wandb/sdk/system/system_monitor.py:97]: clean this one

File: wandb/sdk/system/assets/ipu.py (hunk 6 above)
Comment (dmitryduev) [reply wandb/sdk/system/assets/ipu.py:164]: 👍🏻 done

File: wandb/sdk/system/assets/tpu.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+except (
+    ImportError,
+    TypeError,
+    AttributeError,
+):  # Saw type error when iterating paths on colab...
+    # TODO: Saw error in sentry where module 'tensorflow.python.pywrap_tensorflow'
+    #  has no attribute 'TFE_DEVICE_PLACEMENT_EXPLICIT'
+    tpu_cluster_resolver = None
+    profiler_client = None
+
```
Comment (kptkin) [inline wandb/sdk/system/assets/tpu.py:35]: I changed this file maybe this caused the hang

File: tests/unit_tests/test_system_metrics_tpu.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    assert len(tpu_profiler.samples) == 0
+
+    tpu_profiler = TPUUtilization(service_addr="my-tpu")
+    tpu_profiler._profiler_client = MockProfilerClient()
+    tpu_profiler.sample()
+    assert len(tpu_profiler.samples) == 1
+    assert tpu_profiler.samples[0] == 10.1
+
+
+def test_tpu(test_settings):
```
Comment (kptkin) [inline tests/unit_tests/test_system_metrics_tpu.py:71]: This a new test maybe causing the hang?

Comment (dmitryduev) [reply tests/unit_tests/test_system_metrics_tpu.py:71]: yea, let's skip it too.

File: wandb/sdk/system/assets/tpu.py (hunk 10 above)
Comment (dmitryduev) [reply wandb/sdk/system/assets/tpu.py]: ok, reverted. let's see if it helped :)

**Comment by raubitsj** [comment]
Oops. i made comments in the wrong PR... here they are for reference:
https://github.com/wandb/wandb/pull/4385#pullrequestreview-1148082814

File: wandb/sdk/internal/system/assets/interfaces.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                metric.clear()
+        except Exception as e:
+            wandb.termerror(f"Failed to publish metrics: {e}", repeat=False)
+
+    def start(self) -> None:
+        if self._process is None and not self._shutdown_event.is_set():
+            self._process = threading.Thread(
+                target=self.monitor,
+                daemon=True,
+                name=f"MetricsMonitor::{self.asset_name}",
```
Comment (raubitsj) [inline wandb/sdk/internal/system/assets/interfaces.py:165]: I believe thread names are only 15 characters so all these will display the same.... need to use some shorthand for the naming

**Review by raubitsj — APPROVED** [review]

**Review by raubitsj — APPROVED** [review]
Awesome work!  this is so much nicer to build on!

File: wandb/sdk/internal/system/assets/interfaces.py (hunk 12 above)
Comment (kptkin) [reply wandb/sdk/internal/system/assets/interfaces.py:165]: Ah good to know... will fix this...

**Pull request merged**